Time-Dependent-Coupled-Cluster/td-ccsd.py

```diff
@@ -0,0 +1,798 @@
+
+### Compute MP2 in MO basis set to make sure the transformation was correct
+MP2corr_E = np.einsum('ijab,ijab->', MOijab, t2) / 4
+MP2_E = SCF_E + MP2corr_E
+
+print('MO based MP2 correlation energy: %.8f' % MP2corr_E)
+print('MP2 total energy:       %.8f' % MP2_E)
+psi4.driver.p4util.compare_values(psi4.energy('mp2'), MP2_E, 6, 'MP2 Energy')
+
+### Start CCSD iterations
+print('\nStarting CCSD iterations')
+ccsd_tstart = time.time()
+CCSDcorr_E_old = 0.0
+for CCSD_iter in range(1, maxiter + 1):
+    ### Build intermediates
+    Fae = build_Fae(t1, t2)
+    Fmi = build_Fmi(t1, t2)
+    Fme = build_Fme(t1, t2)
+
+    Wmnij = build_Wmnij(t1, t2)
+    Wabef = build_Wabef(t1, t2)
+    Wmbej = build_Wmbej(t1, t2)
+
+    #### Build RHS side of t1 equations
+    rhs_T1  = F[o, v].copy()
+    rhs_T1 += np.einsum('ie,ae->ia', t1, Fae)
+    rhs_T1 -= np.einsum('ma,mi->ia', t1, Fmi)
+    rhs_T1 += np.einsum('imae,me->ia', t2, Fme)
+    rhs_T1 -= np.einsum('nf,naif->ia', t1, MO[o, v, o, v])
+    rhs_T1 -= 0.5 * np.einsum('imef,maef->ia', t2, MO[o, v, v, v])
+    rhs_T1 -= 0.5 * np.einsum('mnae,nmei->ia', t2, MO[o, o, v, o])
+
+    ### Build RHS side of t2 equations
+    rhs_T2 = MO[o, o, v, v].copy()
+
+    # P_(ab) t_ijae (F_be - 0.5 t_mb F_me)
+    tmp = Fae - 0.5 * np.einsum('mb,me->be', t1, Fme)
+    Pab = np.einsum('ijae,be->ijab', t2, tmp)
+    rhs_T2 += Pab
+    rhs_T2 -= Pab.swapaxes(2, 3)
+
+    # P_(ij) t_imab (F_mj + 0.5 t_je F_me)
+    tmp = Fmi + 0.5 * np.einsum('je,me->mj', t1, Fme)
+    Pij = np.einsum('imab,mj->ijab', t2, tmp)
+    rhs_T2 -= Pij
+    rhs_T2 += Pij.swapaxes(0, 1)
+
+    tmp_tau = build_tau(t1, t2)
+    rhs_T2 += 0.5 * np.einsum('mnab,mnij->ijab', tmp_tau, Wmnij)
+    rhs_T2 += 0.5 * np.einsum('ijef,abef->ijab', tmp_tau, Wabef)
+
+    # P_(ij) * P_(ab)
+    # (ij - ji) * (ab - ba)
+    # ijab - ijba -jiab + jiba
+    tmp = np.einsum('ie,ma,mbej->ijab', t1, t1, MO[o, v, v, o])
+    Pijab = np.einsum('imae,mbej->ijab', t2, Wmbej)
+    Pijab -= tmp
+
+    rhs_T2 += Pijab
+    rhs_T2 -= Pijab.swapaxes(2, 3)
+    rhs_T2 -= Pijab.swapaxes(0, 1)
+    rhs_T2 += Pijab.swapaxes(0, 1).swapaxes(2, 3)
+
+    Pij = np.einsum('ie,abej->ijab', t1, MO[v, v, v, o])
+    rhs_T2 += Pij
+    rhs_T2 -= Pij.swapaxes(0, 1)
+
+    Pab = np.einsum('ma,mbij->ijab', t1, MO[o, v, o, o])
+    rhs_T2 -= Pab
+    rhs_T2 += Pab.swapaxes(2, 3)
+
+    ### Update t1 and t2 amplitudes
+    t1 = rhs_T1 / Dia
+    t2 = rhs_T2 / Dijab
+
+    ### Compute CCSD correlation energy
+    CCSDcorr_E = np.einsum('ia,ia->', F[o, v], t1)
+    CCSDcorr_E += 0.25 * np.einsum('ijab,ijab->', MO[o, o, v, v], t2)
+    CCSDcorr_E += 0.5 * np.einsum('ijab,ia,jb->', MO[o, o, v, v], t1, t1)
+
+    ### Print CCSD correlation energy
+    print('CCSD Iteration %3d: CCSD correlation = %3.12f  '\
+          'dE = %3.5E' % (CCSD_iter, CCSDcorr_E, (CCSDcorr_E - CCSDcorr_E_old)))
+    if (abs(CCSDcorr_E - CCSDcorr_E_old) < E_conv):
+        break
+
+    CCSDcorr_E_old = CCSDcorr_E
```

Comment:
Can you steal all of the above from `helper_CCSD.py` like the `CCSD_T.py` code does?